packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -111,14 +129,60 @@ const FreehandLineModel = types
     },
 
     initialDrag({ x, y }) {
+      // DRAG PATH FROM START
       const dist = self.getDistance(x, y, self.lastPoint.x, self.lastPoint.y)
-      if (dist > LINE_RESOLUTION) {
+      if (dist > LINE_RESOLUTION && ((self.isValid && !self.isClosed) || !self.isValid)) {
+        let action = self.drawActions[self.drawActions.length - 1]
+        action.points.push({ 0: x, 1: y });
         self.pathX.push(x)
         self.pathY.push(y)
       }
+      if (self.isValid && self.isClosed) {
+        self.clipPath = []
+      }
+    },
+
+    finish({ x, y }) {
+      console.log('FINISH()')
+      // DRAG PATH END START
+      self.finished = true
+
+      if (!self.isValid) {
+        self.drawActions.pop();
+        self.tool.deleteMark(self)
+      }
+    },
+
+    close() {
+      let {x, y} = self.initialPoint
+      console.log('close()', x, y)
+      self.drawActions.push({
+        type: 'end',
+        points: [{ 0: x, 1: y }],
+        pointIndex: -1,
+        isReversed: false
+      });
+      self.pathX.push(x)
+      self.pathY.push(y)
+      self.dragPoint = null
+      self.targetPoint = null
+      self.clipPath = []
+	  },
+
+    initialize(points) {
+      if (!points) return;
+
+      self.pathX = points.map(({ x, y }) => x)
+      self.pathY = points.map(({ x, y }) => y)
+      self.dragPoint = null
+      self.targetPoint = null
+      self.originalPath = [];
+      self.clipPath = [];
     },
```

Comment:
This method isn't really needed. MST lets you pass in a snapshot when you create model. The new model's properties are populated from the snapshot. 

https://mobx-state-tree.js.org/concepts/snapshots

Reply:
With drawing tools, you can pass a snapshot of a mark, including coordinates in the snapshot, when you create a new mark. The new mark will then be populated with the coordinates from the snapshot. We do this in the stories and tests to create mock marks. 

```js
const mark = tool.createMark(snapshot)
```

Reply:
The API for tools is described in the drawing tools Readme. 
https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/plugins/drawingTools/README.md

Reply:
This didn't quite work for the same reasons as snapshots weren't quite useful for managing undo/redo state. In the later revision of this code some custom logic is incorporated for constructing local model state that helps with rendering/manipulating the model.